keps/sig-auth/20190711-external-credential-providers.md

```diff
@@ -0,0 +1,243 @@
+systems to help with key rotation and protect against key exfiltration.
+
+### Goals
+
+1. Credential rotation without client restart.
+1. Support standard key management solutions.
+1. Support standard token-based protocols.
+1. Provisioning logic lives outside of Kubernetes codebase.
+1. Kubernetes interface is vendor-neutral.
+
+### Non-Goals
+
+1. Exfiltration protection built into Kubernetes.
+1. Kubernetes triggering rotation.
+1. Deprecation of existing authentication options.
+
+## Proposal
+
+A new authentication flow in libraries around `kubeconfig` based on
+executables. Before performing a request, client executes a binary and uses its
+output for authentication.
+
+There are 2 modes of authentication:
+
+1. bearer tokens
+1. mTLS
+
+Provider response is cached and reused in future requests.
+
+Client is configured with a binary path, optional arguments and environment
+variables to pass to it.
+
+### Provider configuration
+
+Configuration is provided via users section of `kubeconfig` file:
+
+```
+apiVersion: v1
+kind: Config
+users:
+- name: my-user
+  user:
+    exec:
+      # API version to use when decoding the ExecCredentials resource. Required.
+      apiVersion: "client.authentication.k8s.io/v1beta1"
+
+      # Command to execute. Required.
+      command: "example-client-go-exec-plugin"
+
+      # Arguments to pass when executing the plugin. Optional.
+      args:
+      - "arg1"
+      - "arg2"
+
+      # Environment variables to set when executing the plugin. Optional.
+      env:
+      - name: "FOO"
+        value: "bar"
+clusters:
+- name: my-cluster
+  cluster:
+    server: "https://1.2.3.4:8080"
+    certificate-authority: "/etc/kubernetes/ca.pem"
+contexts:
+- name: my-cluster
+  context:
+    cluster: my-cluster
+    user: my-user
+current-context: my-cluster
+```
+
+`apiVersion` specifies the expected version of this API that the plugin
+implements. If the version differs, client must return an error.
+
+`command` specifies the path to the provider binary. The file at this path must
+be readable and executable by the client process.
+
+`args` specifies extra arguments passed to the executable.
+
+`env` specifies environment variables to pass to the provider. The environment
+variables set in the client process are not passed.
+
+### Provider input format
+
+```
+{
+  "apiVersion": "client.authentication.k8s.io/v1",
+  "kind": "ExecCredential"
+}
+```
+
+Provider can safely ignore `stdin` since input object doesn't carry any data.
+
+### Provider output format
+
+```
+{
+  "apiVersion": "client.authentication.k8s.io/v1",
+  "kind": "ExecCredential",
+  "status": {
+    "expirationTimestamp": "$EXPIRATION",
+    "token": "$BEARER_TOKEN",
+    "clientKeyData": "$CLIENT_PRIVATE_KEY",
+    "clientCertificateData": "$CLIENT_CERTIFICATE",
+  }
+}
+```
+
+`EXPIRATION` contains the RFC3339 timestamp with credential expiry. This field
+is not set in `handshaker` mode. Client can cache provided credentials until
```

Comment:
handshaker mode is not defined in this doc?

Reply:
Oops, forgot to clean up